hoomd/md/methods/thermostats.py

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) 2009-2022 The Regents of the University of Michigan.
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+"""Provide classes for thermostatting simulations.
+
+Classes are for use with `hoomd.md.methods.ConstantVolume` and
+`hoomd.md.methods.ConstantPressure`.
+
+Important:
+    Ensure that your initial condition includes non-zero particle velocities and
+    angular momenta (when appropriate). The coupling between the thermostat and
+    the velocities / angular momenta occurs via multiplication, so the
+    thermostat cannot convert a zero velocity into a non-zero one except
+    through particle collisions.
+"""
+
+from hoomd.md import _md
+import hoomd
+from hoomd.operation import _HOOMDBaseObject
+from hoomd.data.parameterdicts import ParameterDict
+from hoomd.variant import Variant
+
+
+class Thermostat(_HOOMDBaseObject):
+    """Base thermostat object class.
+
+    Note:
+        Users should use the subclasses and not instantiate `Thermostat`
+        directly.
+    """
+    _remove_for_pickling = _HOOMDBaseObject._remove_for_pickling + ("_thermo",)
+    _skip_for_equality = _HOOMDBaseObject._skip_for_equality | {
+        "_thermo",
+    }
+
+    def __init__(self, kT):
+        param_dict = ParameterDict(kT=Variant)
+        param_dict["kT"] = kT
+        self._param_dict.update(param_dict)
+
+    def _set_thermo(self, filter, thermo):
+        self._filter = filter
+        self._thermo = thermo
+
+
+class MTTK(Thermostat):
+    r"""The Nosé-Hoover thermostat.
+
+    Produces temperature control through a Nosé-Hoover thermostat.
+
+    Args:
+        kT (hoomd.variant.variant_like): Temperature set point for the
+            thermostat :math:`[\mathrm{energy}]`.
+
+        tau (float): Coupling constant for the thermostat
+            :math:`[\mathrm{time}]`
+
+    The translational thermostat has a momentum :math:`\xi` and position
+    :math:`\eta`. The rotational thermostat has momentum
+    :math:`\xi_{\mathrm{rot}}` and position :math:`\eta_\mathrm{rot}`. Access
+    these quantities using `translational_thermostat_dof` and
+    `rotational_thermostat_dof`.
+
+    See Also:
+        `G. J. Martyna, D. J. Tobias, M. L. Klein 1994
+        <http://dx.doi.org/10.1063/1.467468>`_ and `J. Cao, G. J. Martyna 1996
+        <http://dx.doi.org/10.1063/1.470959>`_.
+
+    Attributes:
+        kT (hoomd.variant.variant_like): Temperature set point for the
+            thermostat :math:`[\mathrm{energy}]`.
+
+        tau (float): Coupling constant for the thermostat
+            :math:`[\mathrm{time}]`
+
+        translational_thermostat_dof (tuple[float, float]): Additional degrees
+            of freedom for the translational thermostat (:math:`\xi`,
+            :math:`\eta`)
+
+        rotational_thermostat_dof (tuple[float, float]): Additional degrees
+            of freedom for the rotational thermostat (:math:`\xi_\mathrm{rot}`,
+            :math:`\eta_\mathrm{rot}`)
```

Comment:
we can remove thermostat now.
```suggestion
        translational_dof (tuple[float, float]): Additional degrees
            of freedom for the translational thermostat (:math:`\xi`,
            :math:`\eta`)

        rotational_dof (tuple[float, float]): Additional degrees
            of freedom for the rotational thermostat (:math:`\xi_\mathrm{rot}`,
            :math:`\eta_\mathrm{rot}`)
```